vpp-patches/0015-Controlled-NAT-function.patch

```diff
@@ -7,32 +7,32 @@ Allow to override ED NAT address and port selection in slow path
 by dedicated NAT bindings per user address. Bindings are created via
 UPG plugin and stored in bihash table.
 ---
- src/plugins/nat/in2out_ed.c   | 133 +++++++++++++++++-----
+ src/plugins/nat/in2out_ed.c   | 142 ++++++++++++++++++-----
  src/plugins/nat/nat.c         | 208 +++++++++++++++++++++++++++++++++-
  src/plugins/nat/nat.h         |  44 +++++++
  src/plugins/nat/nat44_cli.c   |  72 ++++++++++++
  src/plugins/nat/nat_format.c  |  17 +++
  src/plugins/nat/nat_inlines.h |  39 +++++++
- 6 files changed, 484 insertions(+), 29 deletions(-)
+ 6 files changed, 492 insertions(+), 30 deletions(-)
 
 diff --git a/src/plugins/nat/in2out_ed.c b/src/plugins/nat/in2out_ed.c
-index 7153a6035..7a7895a72 100644
+index 7153a6035..d70901c9c 100644
 --- a/src/plugins/nat/in2out_ed.c
 +++ b/src/plugins/nat/in2out_ed.c
-@@ -196,6 +196,65 @@ icmp_in2out_ed_slow_path (snat_main_t * sm, vlib_buffer_t * b0,
+@@ -196,6 +196,67 @@ icmp_in2out_ed_slow_path (snat_main_t * sm, vlib_buffer_t * b0,
    return next0;
  }
  
 +
 +static int
-+nat_controlled_alloc_addr_and_port (snat_main_t * sm, u32 rx_fib_index,
-+				    u32 nat_proto, u32 thread_index,
-+				    ip4_address_t r_addr, u16 r_port,
-+				    u8 proto, u16 port_per_thread,
++nat_controlled_alloc_addr_and_port (snat_main_t * sm,
++				    u32 thread_index, ip4_address_t r_addr,
++				    u16 r_port, u8 proto,
 +				    u32 snat_thread_index, snat_session_t * s,
 +				    ip4_address_t * outside_addr,
 +				    u16 * outside_port,
-+				    clib_bihash_kv_16_8_t * out2in_ed_kv)
++				    clib_bihash_kv_16_8_t * out2in_ed_kv,
++				    u32 buf_out_fib_idx)
```

Comment:
The argument ordering is bad. The typical rule in C is to have input arguments before output arguments in functions.
`buf_out_fib_idx` is clearly an input value.